In [ ]:
import pandas as pd
df = pd.read_csv("./data/math.csv")
print(df.head())

    groups  scores
0  group_A      85
1  group_A      88
2  group_A      90
3  group_A      82
4  group_A      87


In [ ]:
# 1. 각 그룹의 수학성적에 대해 샤피로-윌크 검정을 통한 정규성 확인
from scipy import stats

groups = ['group_A', 'group_B', 'group_C', 'group_D']

for group in groups:
    df_g = df[df['groups'] == group]
    statistics, pvalue = stats.shapiro(df_g['scores'])
    print(f'{group}:', '{:.5f}'.format(pvalue))

group_A: 0.90518
group_B: 0.66782
group_C: 0.44733
group_D: 0.25824


0.05 이상이므로 귀무가설(정규성 가지고 있다) 채택

In [30]:
scores = [ df[df['groups'] == group]['scores'] for group in groups ]

statistics, pvalue = stats.levene(scores[0], scores[1], scores[2], scores[3])
print('{:.5f}'.format(pvalue))

0.17270


0.05 이상이므로 귀무가설(등분산성을 가지고 있다) 채택

In [31]:
import statsmodels as sm
from statsmodels.formula.api import ols

model = ols('scores~groups', df).fit()

In [32]:
from statsmodels.stats.anova import anova_lm

print(anova_lm(model))

            df  sum_sq     mean_sq          F        PR(>F)
groups     3.0   411.8  137.266667  34.174274  1.240642e-10
Residual  36.0   144.6    4.016667        NaN           NaN


3. 귀무가설 기각
4. 그룹 변수 자유도: 3
5. 잔차 자유도: 36
6. 성적 제곱합: 411.8
7. 성적 평균제곱: 137.266667
8. F-통계량: 34.174247
9. 성적에 대한 p-value: 1.240642e-10